src/appleseed.studio/mainwindow/rendering/materialdrophandler.cpp

```diff
@@ -0,0 +1,156 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+// Interface header.
+#include "materialdrophandler.h"
+
+// appleseed.studio headers.
+#include "mainwindow/rendering/renderingmanager.h"
+
+// appleseed.renderer headers.
+#include "renderer/kernel/rendering/scenepicker.h"
+
+// Standard headers.
+#include <string>
+
+namespace appleseed {
+namespace studio {
+
+MaterialDropHandler::MaterialDropHandler(
+    const renderer::Project&    project,
+    RenderingManager&           rendering_manager)
+  : m_project(project), m_rendering_manager(rendering_manager)
+{
+    m_selection_menu.reset(new QMenu());
+    connect(m_selection_menu->addAction("Front Side"), SIGNAL(triggered()), SLOT(slot_apply_to_front()));
+    connect(m_selection_menu->addAction("Back Side"), SIGNAL(triggered()), SLOT(slot_apply_to_back()));
+    connect(m_selection_menu->addAction("Both Sides"), SIGNAL(triggered()), SLOT(slot_apply_to_both()));
+}
+
+void MaterialDropHandler::slot_material_dropped(
+    const foundation::Vector2d& drop_pos,
+    const std::string&          material_name)
+{
+    m_drop_pos = drop_pos;
+    m_material_name = material_name;
+    m_selection_menu->exec(QCursor::pos());
+}
+
+namespace
+{
+    struct MaterialSlot
+    {
+        std::string                     m_name;
+        renderer::ObjectInstance::Side  m_side;
+    };
+
+    class ApplyDroppedMaterial
+        : public RenderingManager::IScheduledAction
+    {
+      public:
+        ApplyDroppedMaterial(
+            renderer::ObjectInstance&           instance,
+            const std::vector<MaterialSlot>&    material_slots,
+            const std::string&                  material_name)
+            : m_instance(instance),
+            m_material_slots(material_slots),
+            m_material_name(material_name)
+        {
+        }
+
+        void operator()(renderer::Project& project) override
+        {
+            for (const MaterialSlot &slot : m_material_slots)
+            {
+                m_instance.assign_material(slot.m_name.c_str(), slot.m_side, m_material_name.c_str());
+            }
+        }
+
+      private:
+        renderer::ObjectInstance&       m_instance;
+        const std::vector<MaterialSlot> m_material_slots;
+        const std::string               m_material_name;
+    };
+}
+
+void MaterialDropHandler::assign_material(const renderer::ObjectInstance::Side side)
+{
+    const renderer::ScenePicker scene_picker(m_project);
+    const renderer::ScenePicker::PickingResult result = scene_picker.pick(m_drop_pos);
+    renderer::ObjectInstance* instance = const_cast<renderer::ObjectInstance*>(result.m_object_instance);
+
+    std::vector<MaterialSlot> material_slots;
+    if (side == renderer::ObjectInstance::Side::FrontSide
+        || side == renderer::ObjectInstance::Side::BothSides)
```

Comment:
You can improve readability by moving the `||` at the end of the previous line:

    if (side == renderer::ObjectInstance::Side::FrontSide ||
        side == renderer::ObjectInstance::Side::BothSides)

However, you'll notice that `renderer::ObjectInstance::Side` is actually a bitmask, so you can simply do something like

    if (side & renderer::ObjectInstance::Side::FrontSide)

And the same thing for the backside.